src/graph.py

```diff
@@ -34,14 +34,15 @@ def add_node(self, n):
                         )
                     raise
 
-    def add_edge(self, n1, n2):
+    def add_edge(self, n1, n2, weight):
         """Add a edge from n1 to n2."""
         new_node = self._dict.setdefault(n1, [])
         self._dict.setdefault(n2, [])
-        if (n2) not in new_node:
-            new_node.append(n2)
+        for tup in new_node:
+            if n2 == tup[0]:
+                raise ValueError('This edge already exists.')
         else:
-            raise ValueError('This edge already exists.')
+            new_node.append((n2, weight))
```

Comment:
Another option to consider for constructing nodes is a another dictionary within your `self._dict`. e.g.:
```python
self._dict = {
    node1: {
        node2: 1,
        node3: 9,
    }
    node2: {
        node3: 4
    }
}